File: rd-net/RdFramework/Impl/RdMap.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using JetBrains.Annotations;
using JetBrains.Collections;
using JetBrains.Collections.Synchronized;
using JetBrains.Collections.Viewable;
using JetBrains.Diagnostics;
using JetBrains.Lifetimes;
using JetBrains.Rd.Base;
using JetBrains.Rd.Util;
using JetBrains.Serialization;

// ReSharper disable InconsistentNaming

namespace JetBrains.Rd.Impl
{

  public class RdMap<K, V> : RdReactiveBase, IViewableMap<K, V> where K : notnull
  {
    private readonly ViewableMap<K, V> myMap = new(new SynchronizedDictionary<K, V>()/*to have thread safe print*/);

    public RdMap(CtxReadDelegate<K> readKey, CtxWriteDelegate<K> writeKey, CtxReadDelegate<V> readValue, CtxWriteDelegate<V> writeValue)
    {
      ValueCanBeNull = false;

      ReadKeyDelegate = readKey;
      WriteKeyDelegate = writeKey;
      
      ReadValueDelegate = readValue;
      WriteValueDelegate = writeValue;
    }


    #region Serializers

    [PublicAPI]
    public CtxReadDelegate<K> ReadKeyDelegate { get; private set; }
    [PublicAPI]
    public CtxWriteDelegate<K> WriteKeyDelegate { get; private set; }

    [PublicAPI]
    public CtxReadDelegate<V> ReadValueDelegate { get; private set; }
    [PublicAPI]
    public CtxWriteDelegate<V> WriteValueDelegate { get; private set; }

    [PublicAPI]
    public static RdMap<K, V> Read(SerializationCtx ctx, UnsafeReader reader)
    {
      return Read(ctx, reader, Polymorphic<K>.Read, Polymorphic<K>.Write, Polymorphic<V>.Read, Polymorphic<V>.Write);
    }
    [PublicAPI]
    public static RdMap<K,V> Read(SerializationCtx ctx, UnsafeReader reader, CtxReadDelegate<K> readKey, CtxWriteDelegate<K> writeKey, CtxReadDelegate<V> readValue, CtxWriteDelegate<V> writeValue)
    {
      var id = reader.ReadRdId();
      return new RdMap<K, V>(readKey, writeKey, readValue, writeValue).WithId(id);
    }

    [PublicAPI]
    public static void Write(SerializationCtx ctx, UnsafeWriter writer, RdMap<K, V> value)
    {
      Assertion.Assert(!value.RdId.IsNil);
      writer.Write(value.RdId);
    }
    #endregion


    #region Versions

    private const int versionedFlagShift = 8;
    private const int Ack = (int)AddUpdateRemove.Remove + 1;
    
    public bool IsMaster = false;
    private long myNextVersion;
    private readonly Dictionary<K, long> myPendingForAck = new Dictionary<K, long>();

    #endregion


    #region Init
    
    public bool OptimizeNested { [PublicAPI] get; set; }
    private volatile SynchronizedDictionary<K, LifetimeDefinition?>? myBindDefinitions;
    
    protected override void PreInit(Lifetime lifetime, IProtocol proto)
    {
      base.PreInit(lifetime, proto);

      if (!OptimizeNested)
      {
        var definitions  = new SynchronizedDictionary<K, LifetimeDefinition?>(myMap.Count); 
        
        foreach (var (key, value) in this)
        {
          if (value != null)
          {
            value.IdentifyPolymorphic(proto.Identities, proto.Identities.Next(RdId));
            var definition = TryPreBindValue(lifetime, key, value, false);
            if (definition != null)
              definitions.Add(key, definition);
          }
        }

        using var cookie = lifetime.UsingExecuteIfAlive();
        if (cookie.Succeed)
        {
          Assertion.Assert(myBindDefinitions == null);
          myBindDefinitions = definitions;
        }
        else
          return;
      }

      proto.Wire.Advise(lifetime, this);
    }

    protected override void Init(Lifetime lifetime, IProtocol proto, SerializationCtx ctx)
    {
      base.Init(lifetime, proto, ctx);
      
      if (!OptimizeNested)
      {
        Change.Advise(lifetime, it =>
        {
          AssertNullability(it.Key);

          if (it.Kind != AddUpdateRemove.Remove) AssertNullability(it.NewValue);

          if (IsLocalChange)
          {
            var definitions = TryGetBindDefinitions(lifetime);
            if (definitions == null)
              return;

            if (it.Kind != AddUpdateRemove.Add) 
              definitions[it.Key]?.Terminate();

            if (it.Kind == AddUpdateRemove.Remove)
              definitions.Remove(it.Key);

            if (it.Kind != AddUpdateRemove.Remove)
            {
              it.NewValue.IdentifyPolymorphic(proto.Identities, proto.Identities.Next(RdId));
              var definition = TryPreBindValue(lifetime, it.Key, it.NewValue, false);
              definitions[it.Key] = definition;
            }
          }
        });
      }

      using (UsingLocalChange())
      {
        Advise(lifetime, it =>
        {
          if (!IsLocalChange) return;

          proto.Wire.Send(RdId, SendContext.Of(ctx, it, this), static (sendContext, stream) =>
          {
            var sContext = sendContext.SzrCtx;
            var evt = sendContext.Event;
            var me = sendContext.This;
            var versionedFlag = me.IsMaster ? 1 << versionedFlagShift : 0;
            stream.Write(versionedFlag | (int)evt.Kind);

            var version = ++me.myNextVersion;
            if (me.IsMaster)
            {
              lock(me.myPendingForAck)
                me.myPendingForAck[evt.Key] = version;
              
              stream.Write(version);
            }

            me.WriteKeyDelegate(sContext, stream, evt.Key);
            if (evt.IsUpdate || evt.IsAdd)
              me.WriteValueDelegate(sContext, stream, evt.NewValue);


            SendTrace?.Log($"{me} :: {evt.Kind} :: key = {evt.Key.PrintToString()}"
                           + (me.IsMaster ? " :: version = " + version : "")
                           + (evt.Kind != AddUpdateRemove.Remove ? " :: value = " + evt.NewValue.PrintToString() : ""));
          });

          if (!OptimizeNested) 
            it.NewValue.BindPolymorphic();
        });
      }
    }

    protected override void Unbind()
    {
      base.Unbind();
      myBindDefinitions = null;
    }

    public override void OnWireReceived(IProtocol proto, SerializationCtx ctx, UnsafeReader stream, IRdWireableDispatchHelper dispatchHelper)
    {
      var header = stream.ReadInt();
      var msgVersioned = (header >> versionedFlagShift) != 0;
      var opType = header & ((1 << versionedFlagShift) - 1);

      var version = msgVersioned ? stream.ReadLong() : 0L;

      var key = ReadKeyDelegate(ctx, stream);

      var lifetime = dispatchHelper.Lifetime;
      if (opType == Ack)
      {
        dispatchHelper.Dispatch(() =>
        {
          lock (myPendingForAck)
          {
            string? error = null;
            if (!msgVersioned) error = "Received ACK while msg hasn't versioned flag set";
            else if (!IsMaster) error = "Received ACK when not a Master";
            else if (!myPendingForAck.TryGetValue(key, out var pendingVersion)) error = "No pending for ACK";
            else if (pendingVersion < version)
              error = $"Pending version `{pendingVersion}` < ACK version `{version}`";
            // Good scenario
            else if (pendingVersion == version)
              myPendingForAck.Remove(key);
            //else do nothing, silently drop
            var isError = !string.IsNullOrEmpty(error);
            if (ourLogReceived.IsTraceEnabled() || isError)
            {
              ourLogReceived.LogFormat(isError ? LoggingLevel.ERROR : LoggingLevel.TRACE,
                "{0}  :: ACK :: key = {1} :: version = {2}{3}", this, key.PrintToString(), version,
                isError ? " >> " + error : "");
            }
          }
        });
      }
      else
      {
        var kind = (AddUpdateRemove)opType;
        var isPut = kind is AddUpdateRemove.Add or AddUpdateRemove.Update;
        var value = isPut ? ReadValueDelegate(ctx, stream) : default;
        var definition = TryPreBindValue(lifetime, key, value, true);

        dispatchHelper.Dispatch(() =>
        {
          if (msgVersioned || !IsMaster || !IsPendingForAck(key))
          {
            ReceiveTrace?.Log($"{this} :: {kind} :: key = {key.PrintToString()}" +
                              (msgVersioned ? " :: version = " + version : "") +
                              (isPut ? $" :: value = {value.PrintToString()}" : ""));
            
            if (isPut)
            {
              if (TryGetBindDefinitions(lifetime) is { } definitions)
              {
                if (kind == AddUpdateRemove.Update)
                  definitions[key]?.Terminate();

                definitions[key] = definition;
              }

              myMap[key] = value!;
            }
            else
            {
              if (TryGetBindDefinitions(lifetime) is { } definitions && definitions.TryGetValue(key, out var prevDefinition))
              {
                prevDefinition?.Terminate();
                definitions.Remove(key);
              }

              myMap.Remove(key);
            }
          }
          else
          {
            ReceiveTrace?.Log(">> CHANGE IGNORED");
          }

          if (msgVersioned)
          {
            proto.Wire.Send(RdId, innerWriter =>
            {
              innerWriter.Write((1 << versionedFlagShift) | Ack);
              innerWriter.Write(version);
              WriteKeyDelegate.Invoke(ctx, innerWriter, key);

              SendTrace?.Log($"{this} :: ACK :: key = {key.PrintToString()} :: version = {version}");
            });

            if (IsMaster)
              ourLogReceived.Error("Both ends are masters: {0}", Location);
          }
        });
      }
    }

    private SynchronizedDictionary<K, LifetimeDefinition?>? TryGetBindDefinitions(Lifetime lifetime)
    {
      var definitions = myBindDefinitions;
      return lifetime.IsAlive ? definitions : null;
    }

    private bool IsPendingForAck(K key)
    {
      lock (myPendingForAck)
        return myPendingForAck.ContainsKey(key);
    }
    
    private LifetimeDefinition? TryPreBindValue(Lifetime lifetime, K key, V? value,  bool bindAlso)
    {
      if (OptimizeNested || !value.IsBindable())
        return null;
      
      var definition = new LifetimeDefinition { Id = value };
      try
      {
        value.PreBindPolymorphic(definition.Lifetime, this, "["+key+"]");
        if (bindAlso)
          value.BindPolymorphic();
        
        lifetime.Definition.Attach(definition, true);
        return definition;
      }
      catch
      {
        definition.Terminate();
        throw;
      }
    }

    #endregion


    #region Read delegation

    IEnumerator IEnumerable.GetEnumerator()
    {
      return myMap.GetEnumerator();
    }

    public IEnumerator<KeyValuePair<K, V>> GetEnumerator()
    {
      return myMap.GetEnumerator();
    }

    public bool Contains(KeyValuePair<K, V> item)
    {
      return myMap.Contains(item);
    }

    public void CopyTo(KeyValuePair<K, V>[] array, int arrayIndex)
    {
      myMap.CopyTo(array, arrayIndex);
    }


    public int Count => myMap.Count;

    public bool IsReadOnly => myMap.IsReadOnly;

    public bool ContainsKey(K key)
    {
      return myMap.ContainsKey(key);
    }


    public bool TryGetValue(K key, out V value)
    {
      return myMap.TryGetValue(key, out value);
    }

    public ICollection<K> Keys => myMap.Keys;


    public ICollection<V> Values => myMap.Values;


    public ISource<MapEvent<K, V>> Change => myMap.Change;

    #endregion


    #region Write delegation
    //todo async!
    public void Add(K key, V value)
    {
      AssertNullability(value);
      using (UsingLocalChange())
      {
        myMap.Add(key, value);
      }
    }


    public bool Remove(K key)
    {
      using (UsingLocalChange())
      {
        return myMap.Remove(key);
      }
    }


    public V this[K key]
    {
      get => myMap[key];
      set
      {
        AssertNullability(value);
        using (UsingLocalChange())
        {
          myMap[key] = value;
        }
      }
    }


    public bool Remove(KeyValuePair<K, V> item)
    {
      using (UsingLocalChange())
      {
        return myMap.Remove(item);
      }
    }


    public void Add(KeyValuePair<K, V> item)
    {
      AssertNullability(item.Value);
      using (UsingLocalChange())
      {
        myMap.Add(item);
      }
    }


    public void Clear()
    {
      using (UsingLocalChange())
      {
        myMap.Clear();
      }
    }

    #endregion


    public void Advise(Lifetime lifetime, Action<MapEvent<K, V>> handler)
    {
      if (IsBound) AssertThreading();
      using (UsingDebugInfo())
      {
        myMap.Advise(lifetime, handler);
      }
    }

    public override RdBindableBase? FindByRName(RName rName)
    {
      var rootName = rName.GetNonEmptyRoot();
      var localName = rootName.LocalName.ToString();
      if (!localName.StartsWith("[") || !localName.EndsWith("]"))
        return null;

      var stringKey = localName.Substring(1, localName.Length - 2);

      foreach (var (key, value) in myMap)
      {
        if (key.ToString() != stringKey)
          continue;

        if (!(value is RdBindableBase bindableValue))
          break;

        if (rootName == rName)
          return bindableValue;

        return bindableValue.FindByRName(rName.DropNonEmptyRoot());
      }

      return null;
    }

    protected override string ShortName => "map";

    public override void Print(PrettyPrinter printer)
    {
      base.Print(printer);
      if (!printer.PrintContent) return;
      
      printer.Print(" [");
      if (Count > 0) printer.Println();

      using (printer.IndentCookie())
      {
        foreach (var kv in this)
        {
          kv.Key.PrintEx(printer);
          printer.Print(" => ");
          kv.Value.PrintEx(printer);
          printer.Println();
        }
      }
      printer.Println("]");
    }
  }

}

File: rd-kt/rd-framework/src/main/kotlin/com/jetbrains/rd/framework/impl/RdMap.kt
package com.jetbrains.rd.framework.impl

import com.jetbrains.rd.framework.*
import com.jetbrains.rd.framework.base.*
import com.jetbrains.rd.util.*
import com.jetbrains.rd.util.collections.SynchronizedMap
import com.jetbrains.rd.util.lifetime.Lifetime
import com.jetbrains.rd.util.lifetime.LifetimeDefinition
import com.jetbrains.rd.util.lifetime.isAlive
import com.jetbrains.rd.util.reactive.*
import com.jetbrains.rd.util.string.*
import kotlin.assert


@Suppress("UNUSED_PARAMETER")
open class RdMap<K : Any, V : Any> private constructor(
    val keySzr: ISerializer<K>,
    val valSzr: ISerializer<V>,
    private val map: ViewableMap<K, V>
) : RdReactiveBase(), IAsyncViewableMap<K, V>, IMutableViewableMap<K, V> {

    companion object : ISerializer<RdMap<*, *>> {
        private enum class Op {Add, Update, Remove, Ack}

//        override val _type : Class<*> get() = throw IllegalStateException("Mustn't be used for polymorphic marshalling")
        fun<K:Any, V:Any> read(ctx: SerializationCtx, buffer: AbstractBuffer, keySzr: ISerializer<K>, valSzr: ISerializer<V>): RdMap<K, V> = RdMap(keySzr, valSzr).withId(RdId.read(buffer))
        override fun write(ctx: SerializationCtx, buffer: AbstractBuffer, value: RdMap<*, *>) = value.rdid.write(buffer)

        override fun read(ctx: SerializationCtx, buffer: AbstractBuffer): RdMap<*, *> {
            return read(ctx, buffer, Polymorphic<Any>(), Polymorphic<Any>())
        }

        const val versionedFlagShift = 8
    }

    override fun deepClone(): IRdBindable = RdMap(keySzr, valSzr).also { for ((k,v) in map) { it[k] = v.deepClonePolymorphic() } }

    private var nextVersion = 0L
    private val pendingForAck = mutableMapOf<K, Long>()
    private var bindDefinitions: MutableMap<K, LifetimeDefinition?>? = null

    var optimizeNested: Boolean = false
        set(value) {
            field = value
            if (value) master = false
        }

    private fun logmsg(op: Op, version: Long, key: K, value: V? = null) : String {
        return "map `$location` ($rdid) :: ${op.name} :: key = ${key.printToString()}"+
            (version > 0).condstr   { " :: version = $version" }  +
            (value != null).condstr { " :: value = ${value.printToString()}" }
    }

    override fun preInit(lifetime: Lifetime, proto: IProtocol) {
        super.preInit(lifetime, proto)

        if (!optimizeNested) {
            val definitions = SynchronizedMap<K, LifetimeDefinition?>()

            for ((key, value) in this) {
                if (value != null) {
                    value.identifyPolymorphic(proto.identity, proto.identity.next(rdid))
                    val definition = tryPreBindValue(lifetime, key, value, false)
                    if (definition != null)
                        definitions[key] = definition
                }
            }

            lifetime.executeIfAlive {
                assert(bindDefinitions == null)
                bindDefinitions = definitions
            }
        }

        proto.wire.advise(lifetime, this)
    }

    override fun init(lifetime: Lifetime, proto: IProtocol, ctx: SerializationCtx) {
        super.init(lifetime, proto, ctx)

        if (!optimizeNested) {
            change.advise(lifetime) {
                if (isLocalChange) {
                    val definitions = tryGetBindDefinitions(lifetime) ?: return@advise

                    if (it !is IViewableMap.Event.Add)
                        definitions[it.key]?.terminate()

                    if (it !is IViewableMap.Event.Remove) {
                        val value = it.newValueOpt
                        value.identifyPolymorphic(proto.identity, proto.identity.next(rdid))
                        val definition = tryPreBindValue(lifetime, it.key, value, false)
                        definitions.put(it.key, definition)?.terminate()
                    }
                }
            }
        }

        localChange { advise(lifetime) lambda@{
            if (!isLocalChange) return@lambda

            proto.wire.send(rdid) { buffer ->
                val versionedFlag = (if (master) 1 else 0) shl versionedFlagShift
                val op = when (it) {
                    is IViewableMap.Event.Add ->    Op.Add
                    is IViewableMap.Event.Update -> Op.Update
                    is IViewableMap.Event.Remove -> Op.Remove
                }
                buffer.writeInt(versionedFlag or op.ordinal)

                val version = if (master) ++nextVersion else 0L

                if (master) {
                    Sync.lock(pendingForAck) {
                        pendingForAck.put(it.key, version)
                    }

                    buffer.writeLong(version)
                }

                keySzr.write(ctx, buffer, it.key)

                it.newValueOpt?.let { valSzr.write(ctx, buffer, it) }

                logSend.trace { logmsg(op, version, it.key, it.newValueOpt) }
            }

            if (!optimizeNested)
                it.newValueOpt.bindPolymorphic()
        }}
    }

    override fun findByRName(rName: RName): RdBindableBase? {
        if (rName == RName.Empty) return this
        val rootName = rName.getNonEmptyRoot()
        val localName = rootName.localName
        if (!localName.startsWith('[') || !localName.endsWith(']'))
            return null

        val stringKey = localName.removeSurrounding("[", "]")

        val entry = map.entries.find { (key, _) -> key.toString() == stringKey }
        val value = entry?.value as? RdBindableBase 
            ?: return null

        if (rootName == rName)
            return value

        return value.findByRName(rName.dropNonEmptyRoot())
    }

    override fun onWireReceived(proto: IProtocol, buffer: AbstractBuffer, ctx: SerializationCtx, dispatchHelper: IRdWireableDispatchHelper) {
        val header = buffer.readInt()
        val msgVersioned = (header shr versionedFlagShift) != 0
        val op = parseFromOrdinal<Op>(header and ((1 shl versionedFlagShift) - 1))

        val version = if (msgVersioned) buffer.readLong() else 0

        val key = keySzr.read(ctx, buffer)

        if (op == Op.Ack) {
            dispatchHelper.dispatch {
                val errmsg = Sync.lock(pendingForAck) {
                    if (!msgVersioned) "Received ${Op.Ack} while msg hasn't versioned flag set"
                    else if (!master) "Received ${Op.Ack} when not a Master"
                    else pendingForAck[key]?.let { pendingVersion ->
                        if (pendingVersion < version) "Pending version `$pendingVersion` < ${Op.Ack} version `$version`"
                        else {
                            //side effect
                            if (pendingVersion == version) pendingForAck.remove(key) //else we don't need to remove, silently drop
                            "" //return good result
                        }
                    } ?: "No pending for ${Op.Ack}"
                }

                if (errmsg.isEmpty())
                    logReceived.trace { logmsg(Op.Ack, version, key) }
                else
                    logReceived.error { logmsg(Op.Ack, version, key) + " >> $errmsg" }
            }

        } else {
            val isPut = (op == Op.Add || op == Op.Update)
            val value = if (isPut) valSzr.read(ctx, buffer) else null

            val lifetime = dispatchHelper.lifetime
            val definition = tryPreBindValue(lifetime, key, value, true)

            dispatchHelper.dispatch {
                if (msgVersioned || !master || !isPendingForAck(key)) {
                    logReceived.trace { logmsg(op, version, key, value) }

                    if (value != null) {
                        val definitions = tryGetBindDefinitions(lifetime)
                        if (definitions != null) {
                            if (op == Op.Update)
                                definitions[key]?.terminate()

                            definitions[key] = definition
                        }

                        map[key] = value
                    } else {
                        val prevDef = tryGetBindDefinitions(lifetime)?.remove(key)
                        prevDef?.terminate()
                        map.remove(key)
                    }

                } else {
                    logReceived.trace { logmsg(op, version, key, value) + " >> REJECTED" }
                }

                if (msgVersioned) {
                    proto.wire.send(rdid) { innerBuffer ->
                        innerBuffer.writeInt((1 shl versionedFlagShift) or Op.Ack.ordinal)
                        innerBuffer.writeLong(version)
                        keySzr.write(ctx, innerBuffer, key)

                        logSend.trace { logmsg(Op.Ack, version, key) }
                    }

                    if (master) logReceived.error { "Both ends are masters: $location" }
                }
            }
        }
    }

    private fun isPendingForAck(key: K): Boolean {
        Sync.lock(pendingForAck) {
            return pendingForAck.containsKey(key)
        }
    }

    private fun tryGetBindDefinitions(lifetime: Lifetime): MutableMap<K, LifetimeDefinition?>?  {
        val definitions = bindDefinitions
        return if (lifetime.isAlive) definitions else null
    }

    private fun tryPreBindValue(lifetime: Lifetime, key: K, value: V?, bindAlso: Boolean): LifetimeDefinition? {
        if (optimizeNested || value == null)
            return null

        val definition = LifetimeDefinition().apply { id = value }
        try {
            value.preBindPolymorphic(definition.lifetime, this, "[$key]")
            if (bindAlso)
                value.bindPolymorphic()

            (lifetime as LifetimeDefinition).attach(definition, true)

            return definition
        } catch (e: Throwable) {
            definition.terminate()
            throw e
        }
    }


    constructor(keySzr: ISerializer<K> = Polymorphic(), valSzr: ISerializer<V> = Polymorphic()) : this(keySzr, valSzr, ViewableMap(SynchronizedMap() /*to have thread=safe print*/))



    override fun print(printer: PrettyPrinter) {
        super.print(printer)
        printer.print(" [")
        if (!isEmpty()) printer.println()

        printer.indent {
            forEach {
                val (key, value) = it
                key.print(printer)
                printer.print(" => ")
                value.print(printer)
                printer.println()
            }
        }
        printer.print("]")
    }

    override val size: Int
        get() = map.size

    override fun containsKey(key: K): Boolean = map.containsKey(key)

    override fun containsValue(value: V): Boolean = map.containsValue(value)

    override fun get(key: K): V? = map.get(key)

    override fun isEmpty(): Boolean = map.isEmpty()

    override val entries
        get() = map.entries
    override val keys
        get() = map.keys
    override val values
        get() = map.values

    override val change: ISource<IViewableMap.Event<K, V>>
        get() = map.change

    override fun put(key: K, value: V): V? = localChange { map.put(key, value) }
    override fun remove(key: K): V? = localChange { map.remove(key) }
    override fun remove(key: K, value: V): Boolean = localChange {
        if (this[key] == value) {
            remove(key)
            return@localChange true
        } else {
            return@localChange false
        }
    }

    override fun clear() = localChange { map.clear() }
    override fun putAll(from: Map<out K, V>) = localChange { map.putAll(from) }


    override fun advise(lifetime: Lifetime, handler: (IViewableMap.Event<K, V>) -> Unit) {
        if (isBound) assertThreading()
        map.advise(lifetime, handler)
    }

    override fun adviseOn(lifetime: Lifetime, scheduler: IScheduler, handler: (IViewableMap.Event<K, V>) -> Unit) {
        if (isBound) assertThreading()
        map.advise(lifetime) { e -> scheduler.invokeOrQueue { handler(e) }}
    }
}
